### DISCUSSION TITLE
Proper use of callback functions

### DISCUSSION THREAD

**Discussion by NageshEranki** [bitcraze/discussions#1011] [Q&A] [answered]
Hi

As I understand, the Bitcraze platform relies on a system of callback functions for different subsystems (logging, parameters, localization, etc). This is reflected in the Crazyflie firmware and the python library as well. I'm trying to write custom code for my research so I need to register my custom callback functions for the various subsystems. However, I don't know what arguments the registered callback function must accept. 

The snippet below illustrates my question: what should the arguments `args1, args2` be?
```
def some_function(args1, args2):
   # Do something

cf = Crazyflie()
cf.connected.add_callback(some_function)
```

**Comment by knmcguire** [comment]
Hi! Have you tried out the [step by step tutorial](https://www.bitcraze.io/documentation/tutorials/#software)s yet, like this one for parameters and logging? https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_connect_log_param/

**Reply by NageshEranki** [reply]
Hi

I went through the tutorials. However, I still need a way for me know what my function signature should look like when I register it as a callback for some subsystem. Taking the following example from your docs:

```
def simple_param_async(scf, groupstr, namestr):
    cf = scf.cf
    full_name = groupstr + '.' + namestr
    cf.param.add_update_callback(group=groupstr,name=namestr,cb=param_stab_est_callback)
    time.sleep(1)
```

Your particular example tells me that callback `param_stab_est_callback()` should accept arguments `name, value`. That is fine, but I don't know a-priori which arguments it should accept. This interfacing of your library functions with my custom functions is a challenge. Is there a trick to do this easier?

Thanks

**Reply by ataffanel** [reply, marked as answer]
Unfortunately the best you can do right now is to look at examples or to look at the source code for what calls the caller you put a callback on. This is a limitation of the type-less python the lib is using and a lack of documentation of the lib.

For the future we hope to switch to type-hint everywhere it is possible to allow editors to help with this kind of things, we will also most likely reorganize the structure of the lib since callbacks are not so user-friendly (I am thinking generators and/or asyncio functions, this is still in the study phase, but it should make things look much less magic/strange).
